cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,710 @@
+  const rmm::device_uvector<typename GraphViewType::edge_type>& global_out_degrees)
+{
+  using vertex_t    = typename GraphViewType::vertex_type;
+  using edge_t      = typename GraphViewType::edge_type;
+  using partition_t = matrix_partition_device_view_t<typename GraphViewType::vertex_type,
+                                                     typename GraphViewType::edge_type,
+                                                     typename GraphViewType::weight_type,
+                                                     GraphViewType::is_multi_gpu>;
+  rmm::device_uvector<edge_t> active_source_degrees(active_sources.size(), handle.get_stream());
+
+  std::vector<vertex_t> id_begin;
+  std::vector<vertex_t> id_end;
+  std::vector<vertex_t> count_offsets;
+  id_begin.reserve(graph_view.get_number_of_local_adj_matrix_partitions());
+  id_end.reserve(graph_view.get_number_of_local_adj_matrix_partitions());
+  count_offsets.reserve(graph_view.get_number_of_local_adj_matrix_partitions());
+  vertex_t counter{0};
+  for (size_t i = 0; i < graph_view.get_number_of_local_adj_matrix_partitions(); ++i) {
+    auto matrix_partition = partition_t(graph_view.get_matrix_partition_view(i));
+    // Starting vertex ids of each partition
+    id_begin.push_back(graph_view.get_local_adj_matrix_partition_row_first(i));
+    id_end.push_back(graph_view.get_local_adj_matrix_partition_row_last(i));
+    count_offsets.push_back(counter);
+    counter += matrix_partition.get_major_size();
+  }
+  rmm::device_uvector<vertex_t> vertex_id_begin(id_begin.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t> vertex_id_end(id_end.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t> vertex_count_offsets(count_offsets.size(), handle.get_stream());
+  raft::update_device(
+    vertex_id_begin.data(), id_begin.data(), id_begin.size(), handle.get_stream());
+  raft::update_device(vertex_id_end.data(), id_end.data(), id_end.size(), handle.get_stream());
+  raft::update_device(
+    vertex_count_offsets.data(), count_offsets.data(), count_offsets.size(), handle.get_stream());
+
+  auto& comm           = handle.get_comms();
+  auto const comm_rank = comm.get_rank();
+
+  thrust::transform(handle.get_thrust_policy(),
+                    active_sources.begin(),
+                    active_sources.end(),
+                    active_source_degrees.begin(),
+                    [id_begin = vertex_id_begin.data(),
+                     id_end   = vertex_id_end.data(),
+                     comm_rank,
+                     global_out_degrees   = global_out_degrees.data(),
+                     vertex_count_offsets = vertex_count_offsets.data(),
+                     count                = vertex_id_end.size()] __device__(auto v) {
+                      // Find which partition id did the source belong to
+                      auto partition_id = thrust::distance(
+                        id_end, thrust::upper_bound(thrust::seq, id_end, id_end + count, v));
```

Comment:
And if `active_sources` are sorted, you really don't need to run thrust::upper_bound for every element. But there is sorting overhead instead.... So, not sure which is better in small-scale, but in large-scale, see the discussion below.

And if I am not mistaken, you're getting `global_out_degrees` (or I'd better say `global_major_degrees`) for the entire major range (which is V/P * col_comm_size). Beware that this can be a memory capacity bottleneck if col_comm_size gets large.

What you can do instead is loop over matrix_partitions (or edge_partitions as we have discussed about improving the primitives API https://github.com/rapidsai/cugraph/issues/2003), broadcast global degrees for the major range, and update global degrees for the sources that belong to the matrix partition range.

And here, we can manage memory footprint vs parallelism trade-offs by processing multiple edge partitions concurrently using multiple streams.

You may not need to do this from the beginning, but if you really want to go to very large scale, you may consider making this updates at some point.

Reply:
We will be getting a sense of the workload when this is used in sampling and I will be inclined to make changes to address those bottlenecks.